### PR TITLE
Use proper classloader when loading service with ServiceLoader

### PR DESCRIPTION
**Pull request by ctomc**
+ add option to pass custom classloader
+ includes PRs #22 #21 #20 
+ some build cleanup & use wildfly checkstyle
- moves build requirement to JDK7

I would suggest a version bump

### CODE REVIEW

File: src/main/java/org/jboss/common/beans/property/finder/PropertyEditorFinder.java [hunk 1]
```diff
@@ -86,22 +80,26 @@ private static void init() {
...
+                    init(classLoader);
                     result = _INSTANCE;
                 }
             }
         }
         return result;
     }
+
+    public static PropertyEditorFinder getInstance() {
+        return getInstance(PropertyEditorFinder.class.getClassLoader());
```
Comment (bstansberry) [inline src/main/java/org/jboss/common/beans/property/finder/PropertyEditorFinder.java]: To be compatible this needs to use a doPrivileged call to get the TCCL and then pass that. Per the ServiceLoader javadoc, the previous call at L81 resulted in use of the TCCL, not the calling class' classloader.

Comment (bstansberry) [inline src/main/java/org/jboss/common/beans/property/finder/PropertyEditorFinder.java]: Please ignore that bit about a doPrivileged call. This is a utility lib and shouldn't be making privileged calls like that. The old code (inside ServiceLoader.load(Class)) didn't use a privileged block to access the TCCL.

**Comment by bstansberry** [comment]
This lib (v 1.1.0) is used in EAP so any change like this can't happen in 1.1.1.

BTW, seems I can't merge this anyway.

**Comment by wolfc** [comment]
I don't like the aggregation of the PRs. ~~Nor can I find the associated issues anywhere.~~ The issues are at https://github.com/baranowb/jboss-common-beans/issues which is the wrong place.
@bstansberry comment on TCCL must be resolved, apart from that I don't see any real backwards compatibility issues. At that point a BZ can be filed and we can see whether it can get approved for EAP.

**Comment by ctomc** [comment]
@wolfc there was a long discussion with @bstansberry about the TCCL usage and we agreed that this change is fine.
There was a extra commit to revert behavior to be exactly as it was but was removed after discussion as proposed change fixes broken library which works wrongly in jboss-modules environment as it is.

**Comment by ctomc** [comment]
@wolfc transcript of chat with Brian
```
Tomaz Cerar	@BrianStansberry not sure what you mean here https://github.com/jboss/jboss-common-beans/pull/23 as it looks like your second comment negates first one, so i didn't respond :)	10:01 PM
Brian Stansberry	@ctomc I edited the first one to correct it and removed the 2nd one	10:07 PM
Brian Stansberry	 I originally said it should use a doPrivileged block to get the TCCL	10:07 PM
Brian Stansberry	 it should not use doPrivileged	10:07 PM
Brian Stansberry	 it should just call Thread.currentThread().getContextClassLoader())	10:08 PM
Tomaz Cerar	 hmm, that would make it backward compatible for sure	10:08 PM
Tomaz Cerar	 i dont really want that, as it was broken	10:08 PM
Tomaz Cerar	 and afair only known consument is wildfly/eap at the moment	10:08 PM
Brian Stansberry	 the point of the overloaded method is to let users do it correctly	10:09 PM
Tomaz Cerar	 yeah, and fix the default to be proper	10:09 PM
Tomaz Cerar	 as the ServiceLoader needs to load services that are part of that jar	10:09 PM
Brian Stansberry	 ok, so the idea is to load from the jar by default and require people wanting something else to specify the classloader	10:13 PM
Tomaz Cerar	 yeah	10:13 PM
Tomaz Cerar	 as code is written, it will break if it cannot find META-INF/services that are part of jar itself	10:14 PM
Brian Stansberry	 in a flat classpath that would result in kind of random results	10:14 PM
Tomaz Cerar	 and in flat classpath it works fine with TCCL	10:14 PM
Brian Stansberry	 since presumably the targeted service and the jar's own will both be visible	10:14 PM
Tomaz Cerar	 i dont mind changing to TCCL to mentain compatibilty, but given that wildfly is main user of this lib	10:15 PM
Tomaz Cerar	 it was actually split out of common-core to better support wildfly back in a day	10:15 PM
Tomaz Cerar	 updated PR to use TCCL https://github.com/jboss/jboss-common-beans/pull/23	10:18 PM
Brian Stansberry	 better now. thanks	10:23 PM
Brian Stansberry	 that code was pretty broken in the first place	10:24 PM
Brian Stansberry	 it's kind of the random Service provider	10:24 PM
Brian Stansberry	 with modules it's better as you can ship your own impls and hide the standard ones via no services="import"	10:25 PM
Brian Stansberry	 but with a flat classpath it was probably a crapshoot	10:25 PM
Tomaz Cerar	 true, but currently how it is, that if entry module (org.jboss.as.server) doesn't have deps to org.jboss.common-beans service loader doesnt find anyting	10:25 PM
Tomaz Cerar	 which is a bummer	10:25 PM
Brian Stansberry	 you mean if the deps doesn't have services="import" ?	10:27 PM
Tomaz Cerar	 yup	10:27 PM
Tomaz Cerar	 even better service=export there	10:28 PM
Tomaz Cerar	https://github.com/wildfly/wildfly-core/blob/master/core-feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/server/main/module.xml#L40	10:28 PM
Brian Stansberry	 that's pretty weird	10:29 PM
Tomaz Cerar	 this was done when it was initialy added to the distro, and noone noticed that module itself has no deps whatsoever on common-beans	10:29 PM
Brian Stansberry	 ok then what does org.jboss.as.server have to do with anything?	10:30 PM
Brian Stansberry	 besides someone must be wiring up their module in a stupid way	10:30 PM
Tomaz Cerar	 pretty much every module depends on it	10:30 PM
Tomaz Cerar	 + it is entry point	10:31 PM
Brian Stansberry	 well, that's just broken	10:31 PM
Tomaz Cerar	 yes, that is why i want to have lib fixed ;)	10:31 PM
Tomaz Cerar	 with the changes in the original PR that dep in as.server module ware no longer required	10:32 PM
Brian Stansberry	 sure	10:33 PM
Brian Stansberry	 and the basic point is valid	10:33 PM
Brian Stansberry	 but the dep isn't required regardless; that was just a lazy hack	10:33 PM
Brian Stansberry	 tbh, i understand your original approach better now, and don't really mind	10:34 PM
Brian Stansberry	 although I'm expecting to get some outraged bug report from somewhere on Saturn or something ;)	10:34 PM
Brian Stansberry	 but at least it's a minor version bump, not a micro	10:35 PM
Tomaz Cerar	 yeah that is why i prefer minor bump or even major :)	10:35 PM
Tomaz Cerar	 definitly not micro	10:35 PM
Tomaz Cerar	 and better to get this fixed before EAP7	10:35 PM
Tomaz Cerar	 so let me remove my last commit than	10:35 PM
Tomaz Cerar	 pushed	10:36 PM
Brian Stansberry	 a 3rd option is to leave the TCCL, and add PropertyEditorFinder getLocalInstance()	10:36 PM
Brian Stansberry	 but I don't really care	10:37 PM
Tomaz Cerar	:)	10:37 PM
Tomaz Cerar	 it is a small module and used by really small part of app server	10:37 PM
Tomaz Cerar	 sar subsystem is only conusment
```

**Pull request merged**